src/compile_cache.cc

```diff
@@ -223,13 +226,93 @@ void CompileCacheHandler::ReadCacheFile(CompileCacheEntry* entry) {
   Debug(" success, size=%d\n", total_read);
 }
 
+#ifdef _WIN32
+constexpr bool IsWindowsDeviceRoot(const char c) noexcept {
+  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+}
+#endif
+
+static std::string NormalisePath(std::string_view path) {
+  std::string normalised_string(path);
+  constexpr std::string_view file_scheme = "file://";
+  if (normalised_string.rfind(file_scheme, 0) == 0) {
+    normalised_string.erase(0, file_scheme.size());
+  }
+
+#ifdef _WIN32
+  if (normalised_string.size() > 2 &&
+      IsWindowsDeviceRoot(normalised_string[0]) &&
+      normalised_string[1] == ':' &&
+      (normalised_string[2] == '/' || normalised_string[2] == '\\')) {
```

Comment:
there's a fair amount of duplication with these checks in a couple places. Can this be separated out into a utility method